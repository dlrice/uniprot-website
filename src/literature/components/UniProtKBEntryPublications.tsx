import React, { FC, Fragment } from 'react';
import { connect } from 'react-redux';
import { Publication } from 'franklin-sites';
import { UniProtkbAPIModel } from '../../model/uniprotkb/UniProtkbConverter';
import { RootState } from '../../state/state-types';

const UniProtKBEntryPublications: FC<{
  pubmedIds: string[];
  entryData: UniProtkbAPIModel | null;
}> = ({ pubmedIds, entryData }) => {
  if (!entryData) {
    return null;
  }

  const references =
    entryData.references &&
    entryData.references.filter(
      reference =>
        reference.citation.citationXrefs &&
        reference.citation.citationXrefs.some(
          citation => citation.id && pubmedIds.includes(citation.id)
        )
    );
  return (
    <Fragment>
      {references &&
        references.map(reference => (
          <Publication
            title={reference.citation.title}
            authors={reference.citation.authors}
            key={reference.citation.title}
            journalInfo={{
              firstPage: reference.citation.firstPage,
              journal: reference.citation.journal,
              lastPage: reference.citation.lastPage,
              publicationDate: reference.citation.publicationDate,
              volume: reference.citation.volume,
            }}
          />
        ))}
    </Fragment>
  );
};

const mapStateToProps = (state: RootState) => ({
  entryData: state.entry.data,
});

export default connect(mapStateToProps)(UniProtKBEntryPublications);
