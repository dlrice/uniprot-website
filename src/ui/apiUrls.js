const prefix = '//wwwdev.ebi.ac.uk/uniprot/api';

export default {
  // uniprotkb advanced search terms
  advanced_search_terms: `${prefix}/configure/uniprotkb/search_terms`,
  // Annotation evidence used by advanced search
  annotation_evidences: `${prefix}/configure/uniprotkb/annotation_evidences`,
  // Go evidences used by advanced go search
  // "itemType": "goterm",
  go_evidences: `${prefix}/configure/uniprotkb/go_evidences`,
  // Database cross references used by advanced search
  database_xefs: `${prefix}/configure/uniprotkb/databases`,
  // Database cross reference fields in result column configure
  // "itemType": "database",
  database_fields: `${prefix}/configure/uniprotkb/databasefields`,
  // All result fields except database cross reference fields
  results_fields: `${prefix}/configure/uniprotkb/resultfields`,
};
