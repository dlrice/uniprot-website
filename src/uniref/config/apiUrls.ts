import queryString from 'query-string';
import joinUrl from 'url-join';

export const devPrefix = 'http://wp-np2-be:8091';
export const prodPrefix = 'https://www.ebi.ac.uk';

const apiUrls = {
  search: joinUrl(devPrefix, 'uniprot/api/uniref/search'),
};

export default apiUrls;

export const getAPIQueryUrl = (query: string) => {
  return `${apiUrls.search}?${queryString.stringify({
    query: `${query}`,
  })}`;
};
