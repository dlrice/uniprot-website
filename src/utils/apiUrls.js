import urljoin from 'url-join';
import queryString from 'query-string';

export const joinUrl = (...args) => urljoin(args);

const prefix = '//wwwdev.ebi.ac.uk';

const apiUrls = {
  // uniprotkb advanced search terms
  advanced_search_terms: joinUrl(prefix, '/uniprot/api/configure/uniprotkb/search_terms'),
  // Annotation evidence used by advanced search
  evidences: {
    annotation: joinUrl(prefix, '/uniprot/api/configure/uniprotkb/annotation_evidences'),
    // Go evidences used by advanced go search
    // "itemType": "goterm",
    go: joinUrl(prefix, '/uniprot/api/configure/uniprotkb/go_evidences'),
  },
  // Database cross references used by advanced search
  database_xefs: joinUrl(prefix, '/uniprot/api/configure/uniprotkb/databases'),
  // Database cross reference fields in result column configure
  // "itemType": "database",
  database_fields: joinUrl(prefix, '/uniprot/api/configure/uniprotkb/databasefields'),
  // All result fields except database cross reference fields
  results_fields: joinUrl(prefix, '/uniprot/api/configure/uniprotkb/resultfields'),
  // Retrieve results
  advanced_search: joinUrl(prefix, '/uniprot/api/uniprotkb/search'),
};

export default apiUrls;

const RE_QUERY = /\?$/;

export const getSuggesterUrl = (url, value) => joinUrl(prefix, url.replace(RE_QUERY, value));

export const getUniProtQueryUrl = (encodedUniprotQueryString, columns, filters, cursor) => joinUrl(
  apiUrls.advanced_search,
  '/?',
  queryString.stringify({
    query: encodedUniprotQueryString,
    fields: columns.join(','),
  }),
);
