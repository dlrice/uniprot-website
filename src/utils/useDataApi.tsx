/* eslint-disable @typescript-eslint/no-explicit-any */
import { useState, useEffect } from 'react';
import fetchData from './fetchData';

const useDataApi = (url: string, headers: any) => {
  const [data, setData] = useState<any>({});

  useEffect(() => {
    const getData = async () => {
      const result = await fetchData(url, headers);
      setData(result.data);
    };
    getData();
  }, [url]);

  return data;
};

export default useDataApi;
