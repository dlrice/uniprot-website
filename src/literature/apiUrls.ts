import { joinUrl, devPrefix } from '../utils/apiUrls';

const literatureApiUrls = {
  literature: joinUrl(devPrefix, '/uniprot/api/literature'),
};

export const getPublicationURL = (id: string) =>
  joinUrl(literatureApiUrls.literature, id);

export const getPublicationsURL = (ids: string[]) =>
  `${literatureApiUrls.literature}/searchs?query=(${ids
    .map((id) => `id:${id}`)
    .join(' OR ')})`;

export default literatureApiUrls;
