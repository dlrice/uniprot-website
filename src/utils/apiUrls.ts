import urljoin from 'url-join';
import queryString from 'query-string';
import { v1 } from 'uuid';
import { createEmptyClause } from '../search/utils/clause';

export const joinUrl = (...args: string[]) => urljoin(args);

const prefix = '//wwwdev.ebi.ac.uk';

const apiUrls = {
  // uniprotkb advanced search terms
  advanced_search_terms: joinUrl(prefix, '/uniprot/api/configure/uniprotkb/search_terms'),
  // Annotation evidence used by advanced search
  evidences: {
    annotation: joinUrl(prefix, '/uniprot/api/configure/uniprotkb/annotation_evidences'),
    // Go evidences used by advanced go search
    // "itemType": "goterm",
    go: joinUrl(prefix, '/uniprot/api/configure/uniprotkb/go_evidences'),
  },
  // Database cross references used by advanced search
  database_xefs: joinUrl(prefix, '/uniprot/api/configure/uniprotkb/databases'),
  // Database cross reference fields in result column configure
  // "itemType": "database",
  database_fields: joinUrl(prefix, '/uniprot/api/configure/uniprotkb/databasefields'),
  // All result fields except database cross reference fields
  results_fields: joinUrl(prefix, '/uniprot/api/configure/uniprotkb/resultfields'),
  // Retrieve results
  advanced_search: joinUrl(prefix, '/uniprot/api/uniprotkb/search'),
};

export default apiUrls;

const RE_QUERY = /\?$/;

export const getSuggesterUrl = (url: string, value: string) => joinUrl(prefix, url.replace(RE_QUERY, value));

export const getUniProtQueryUrl = (
  encodedUniprotQueryString: string,
  columns: Array<string>,
  filters: Array<string>,
  cursor: string,
) => `${apiUrls.advanced_search}?${queryString.stringify({
  query: encodedUniprotQueryString,
  fields: columns.join(','),
  includeFacets: true,
})}`;

const findTopLevelParenthesisIndices = (query) => {
  let balance = 0;
  let parenthesisIndices = [0, 0];
  const topLevelParenthesisIndices = [];
  for (let i = 0; i < query.length; i += 1) {
    const c = query[i];
    if (c === '(') {
      balance += 1;
      if (balance === 1) {
        parenthesisIndices = [i, 0];
      }
    } else if (c === ')') {
      balance -= 1;
      if (balance === 0) {
        parenthesisIndices[1] = i;
        topLevelParenthesisIndices.push(parenthesisIndices);
      }
    }
  }
  return topLevelParenthesisIndices;
};

const findSearchTerm = (queryField, value, searchTerms) => searchTerms.find((field) => {
  if (field.items) {
    return findSearchTerm(queryField, value, field.items);
  }
  return field.term === queryField;
});

const ALLOWED_CONJUNCTIONS = ['AND', 'OR', 'NOT'];

const parseClause = (conjunction, fieldValue, searchTerms) => {
  if (!searchTerms) {
    return false;
  }
  const conjunctionUpper = conjunction ? conjunction.toUpperCase() : 'AND';
  const [field, value] = fieldValue.split(':');
  if (!value) {
    const allClause = createEmptyClause();
    allClause.queryInput = { stringValue: field };
    return allClause;
  }
  if (!ALLOWED_CONJUNCTIONS.includes(conjunctionUpper)) {
    throw new Error(`${conjunctionUpper} conjunction is not part of ${ALLOWED_CONJUNCTIONS}`);
  }
  const searchTerm = findSearchTerm(field, value, searchTerms);
  if (!searchTerm) {
    throw new Error(`${field} not a valid field.`);
  }
  const queryInput = {};
  switch (searchTerm.dataType) {
    case 'date':
      // (created:%5B2019-01-01%20TO%200001-12-12%5D)
      console.log(value);
      break;
    case 'enum':
    case 'string':
      if (searchTerm.hasRange) {
        console.log(value);
      }
      queryInput.stringValue = value;
      break;
    case 'integer':
      console.log('integer');
      break;
    default:
      return null;
  }
  return {
    id: v1(),
    logicOperator: conjunctionUpper,
    field: searchTerm,
    queryInput,
  };
};

export const getQueryFromUrl = (query) => {
  const m = query.match(/\?query=(.*)/);
  return m && m[1];
};

export const unpackQueryUrl = (query, searchTerms) => {
  const parenthesisIndices = findTopLevelParenthesisIndices(query);
  let conjunctionIndex = 0;
  if (!parenthesisIndices.length) {
    return [parseClause('AND', query, searchTerms)];
  }
  return parenthesisIndices.map((parenthesis) => {
    const conjunction = query.slice(conjunctionIndex, parenthesis[0]);
    const fieldValue = query.slice(parenthesis[0] + 1, parenthesis[1]);
    conjunctionIndex = parenthesis[1] + 1;
    return parseClause(conjunction, fieldValue, searchTerms);
  });
};
