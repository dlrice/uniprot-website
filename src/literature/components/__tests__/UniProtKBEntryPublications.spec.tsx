import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import UniProtKBEntryPublications from '../UniProtKBEntryPublications';
import { render } from '@testing-library/react';
import { getPublicationsURL } from '../../apiUrls';
import mockUniProtKBEntryPublications from './__mockData__/uniprotKBEntryPublications.json';

import useDataApi from '../../../hooks/useDataApi';
jest.mock('../../../hooks/useDataApi', () => jest.fn());

describe('UniProtKBEntryPublications', () => {
  it('Should make a call with pubmed ids and render properly', () => {
    useDataApi.mockImplementation(() => ({
      loading: false,
      data: mockUniProtKBEntryPublications,
    }));
    const pubMedIds = ['123', '456'];
    const { asFragment } = render(
      <Router>
        <UniProtKBEntryPublications pubmedIds={pubMedIds} />
      </Router>
    );
    expect(useDataApi).toHaveBeenCalledWith(getPublicationsURL(pubMedIds));
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should make a call with pubmed ids', () => {
    useDataApi.mockImplementation(() => ({
      error: {
        message: 'There has been an error',
      },
    }));
    const pubMedIds = ['123', '456'];
    const { asFragment } = render(
      <Router>
        <UniProtKBEntryPublications pubmedIds={pubMedIds} />
      </Router>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
