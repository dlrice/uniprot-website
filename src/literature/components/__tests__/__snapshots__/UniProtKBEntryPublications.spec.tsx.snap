// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UniProtKBEntryPublications Should render 1`] = `
<DocumentFragment>
  <section
    class="publication"
  >
    <h4>
      Leo book tittle
    </h4>
    <section
      class="publication__authors"
    >
      <a
        href="/"
      >
        author Leo
      </a>
    </section>
    <section
      class="publication__columns"
    >
      <section
        class="publication__columns__main"
      />
      <section
        class="publication__columns__side"
      >
        <ul
          class="no-bullet"
        >
          <li>
            <a
              href="//dx.doi.org/undefined"
            >
              journal name undefined:undefined-undefined(date value)
            </a>
          </li>
        </ul>
      </section>
    </section>
  </section>
</DocumentFragment>
`;
